### PR TITLE
[ENH] Add reweighing preprocessing technique based on Kamiran, Calders

### PR DESCRIPTION
**Pull request by wirrywoo**
## Description
This pull request adds a new **reweighing preprocessing technique** based on the method introduced by Kamiran and Calders (2012). The implementation introduces the `KamiranCaldersReweighing` transformer in the `fairlearn.preprocessing` module, enabling users to mitigate bias by assigning instance‑level weights such that the target variable becomes statistically independent of one or more sensitive features.

Key aspects of this PR:
- Full implementation of the Kamiran–Calders reweighing algorithm, including support for multiple sensitive features.
- Encoder support for categorical features and targets, with scikit‑learn–style validation using `validate_data` and `check_is_fitted`.
- Compatibility with NumPy arrays, pandas DataFrames, and **Narwhals DataFrames**, preserving feature names and roundtrip behavior for DataFrame‑like inputs.
- Thorough docstring updates in numpydoc format.
- Updated unit tests that validate probabilistic invariance and handle sensitive feature validation, Narwhals interoperability, and boundary cases.
- Updated user guide `.rst` documentation describing the method, its mathematical foundation, usage examples, and API.

This PR does not modify existing APIs outside of the preprocessing package. Related changes include tests, documentation, and interoperability improvements.

## Tests
- [ ] no new tests required  
- [x] new tests added  
- [ ] existing tests adjusted

This includes:
- Tests for single and multiple sensitive features
- Tests ensuring errors are raised for invalid sensitive features
- Tests ensuring `transform` requires `fit`
- Tests for Narwhals DataFrame roundtrips
- Adjustments of expected statistical invariances in tests

## Documentation
- [x] no documentation changes needed in non‑doc files  
- [x] user guide added or updated  
- [x] API docs added or updated  
- [ ] example notebook added or updated

User guide `.rst` was updated to explain the method, its theoretical basis, and provide an example with a toy dataset. API docstrings were enhanced for clarity and completeness.

Relevant issue addressed: https://github.com/fairlearn/fairlearn/issues/784

### All tests passed:
image

### NumPy Array Support:
```
import numpy as np

df = np.array([
    ("M", "Native",    "H. school", "Board",      "+"),
    ("M", "Native",    "Univ.",     "Board",      "+"),
    ("M", "Native",    "H. school", "Board",      "+"),
    ("M", "Non-nat.",  "H. school", "Healthcare", "+"),
    ("M", "Non-nat.",  "Univ.",     "Healthcare", "-"),
    ("F", "Non-nat.",  "Univ.",     "Education",  "-"),
    ("F", "Native",    "H. school", "Education",  "-"),
    ("F", "Native",    "None",      "Healthcare", "+"),
    ("F", "Non-nat.",  "Univ.",     "Education",  "-"),
    ("F", "Native",    "H. school", "Board",      "+"),
], dtype=object)

X = df[:, :-1]
y = df[:, -1]

reweighing = KamiranCaldersReweighing()

X_with_weight = reweighing.fit_transform(
    X,
    y,
    sensitive_features=[0],  # gender
)

print(X_with_weight)

# Expected Output:
# [['M' 'Native' 'H. school' 'Board' 0.7499999999999999]
#  ['M' 'Native' 'Univ.' 'Board' 0.7499999999999999]
#  ['M' 'Native' 'H. school' 'Board' 0.7499999999999999]
#  ['M' 'Non-nat.' 'H. school' 'Healthcare' 0.7499999999999999]
#  ['M' 'Non-nat.' 'Univ.' 'Healthcare' 2.0]
#  ['F' 'Non-nat.' 'Univ.' 'Education' 0.6666666666666667]
#  ['F' 'Native' 'H. school' 'Education' 0.6666666666666667]
#  ['F' 'Native' 'None' 'Healthcare' 1.4999999999999998]
#  ['F' 'Non-nat.' 'Univ.' 'Education' 0.6666666666666667]
#  ['F' 'Native' 'H. school' 'Board' 1.4999999999999998]]
```

### Pandas DataFrame Support:
```
import pandas as pd

df = pd.DataFrame({
    "Sex": ["M", "M", "M", "M", "M", "F", "F", "F", "F", "F"],
    "Ethnicity": [
        "Native", "Native", "Native", "Non-nat.", "Non-nat.",
        "Non-nat.", "Native", "Native", "Non-nat.", "Native"
    ],
    "Highest degree": [
        "H. school", "Univ.", "H. school", "H. school", "Univ.",
        "Univ.", "H. school", "None", "Univ.", "H. school"
    ],
    "Job type": [
        "Board", "Board", "Board", "Healthcare", "Healthcare",
        "Education", "Education", "Healthcare", "Education", "Board"
    ],
})

y = pd.Series(["+", "+", "+", "+", "-", "-", "-", "+", "-", "+"])

reweighing = KamiranCaldersReweighing()
df_weighted = reweighing.fit_transform(df, y, sensitive_features=["Sex"])

print(df_weighted)

# Expected Output:
#   Sex Ethnicity Highest degree    Job type    weight
# 0   M    Native      H. school       Board  0.750000
# 1   M    Native          Univ.       Board  0.750000
# 2   M    Native      H. school       Board  0.750000
# 3   M  Non-nat.      H. school  Healthcare  0.750000
# 4   M  Non-nat.          Univ.  Healthcare  2.000000
# 5   F  Non-nat.          Univ.   Education  0.666667
# 6   F    Native      H. school   Education  0.666667
# 7   F    Native           None  Healthcare  1.500000
# 8   F  Non-nat.          Univ.   Education  0.666667
# 9   F    Native      H. school       Board  1.500000
```

### Narwhal DataFrame Support:
```
import narwhals.stable.v1 as nw
import pandas as pd

# Original pandas DataFrame
pdf = pd.DataFrame({
    "Sex": ["M", "M", "M", "M", "M", "F", "F", "F", "F", "F"],
    "Ethnicity": [
        "Native", "Native", "Native", "Non-nat.", "Non-nat.",
        "Non-nat.", "Native", "Native", "Non-nat.", "Native"
    ],
    "Highest degree": [
        "H. school", "Univ.", "H. school", "H. school", "Univ.",
        "Univ.", "H. school", "None", "Univ.", "H. school"
    ],
    "Job type": [
        "Board", "Board", "Board", "Healthcare", "Healthcare",
        "Education", "Education", "Healthcare", "Education", "Board"
    ],
    "Cl.": ["+", "+", "+", "+", "-", "-", "-", "+", "-", "+"]
})

# Convert to Narwhals
df = nw.from_native(pdf)
df_features = df.drop(*["Cl."])

# Apply Kamiran-Calders reweighing
reweighing = KamiranCaldersReweighing()
df_weighted = reweighing.fit_transform(df_features, pdf["Cl."], sensitive_features=["Sex"])

print(df_weighted)

# Expected Output:
# ┌────────────────────────────────────────────────────┐
# |                 Narwhals DataFrame                 |
# |----------------------------------------------------|
# |  Sex Ethnicity Highest degree    Job type    weight|
# |0   M    Native      H. school       Board      0.75|
# |1   M    Native          Univ.       Board      0.75|
# |2   M    Native      H. school       Board      0.75|
# |3   M  Non-nat.      H. school  Healthcare      0.75|
# |4   M  Non-nat.          Univ.  Healthcare       2.0|
# |5   F  Non-nat.          Univ.   Education  0.666667|
# |6   F    Native      H. school   Education  0.666667|
# |7   F    Native           None  Healthcare       1.5|
# |8   F  Non-nat.          Univ.   Education  0.666667|
# |9   F    Native      H. school       Board       1.5|
# └────────────────────────────────────────────────────┘
```

### CODE REVIEW

**Comment by wirrywoo** [comment]
Hello all, any updates on this PR review? Thanks!

**Comment by taharallouche** [comment]
> Hello all, any updates on this PR review? Thanks!

I haven't had time to review it yet @wirrywoo i'll do my best

File: fairlearn/preprocessing/_kamiran_calders_reweighing.py [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+            df_out["weight"] = weights
+            return df_out
+        elif self._is_nw_:
+            return nw.from_numpy(
+                np.column_stack([X_valid, weights]),
+                schema=list(self.feature_names_in_) + ["weight"],
+                backend="pandas",
+            )
+        else:
+            return np.column_stack([X_valid, weights])
```
Comment (taharallouche) [inline fairlearn/preprocessing/_kamiran_calders_reweighing.py]:
```suggestion
        check_is_fitted(self)

        X_valid = validate_data(self, X, reset=False, ensure_2d=True, dtype=None)
        n_samples = X_valid.shape[0]

        X_enc = self._x_encoder_.transform(X_valid) if self._x_encoder_ else X_valid

        if self._weights_mapping_ is None:
            weights = np.ones(n_samples)
            return np.column_stack([X_valid, weights])

        y_array = np.asarray(y).ravel()
        if self._y_encoder_:
            y_array = self._y_encoder_.transform(y_array.reshape(-1, 1)).ravel()
        S = X_enc[:, self.sensitive_idx_]
        structured = np.column_stack([S, y_array])
        weights = np.array([self._weights_mapping_.get(tuple(row), 1.0) for row in structured])

        return (
            nw.from_native(X, eager_only=True, pass_through=False)
            .with_columns(weight=weights)
            .to_native()
        )
```

I think that it would look like this with narwhals

File: test/unit/preprocessing/kamiran_calders_reweighing/test_kamiran_calders_reweighing.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+            p_sy,
+            p_s[s] * p_y[y_val],
+            rel_tol=1e-6,
+        )
+
+
+# ----------------------------------------------------------------------
+# 3. Multiple sensitive features
+# ----------------------------------------------------------------------
+def test_multiple_sensitive_features_supported(kamiran_example_dataframe):
```
Comment (taharallouche) [inline test/unit/preprocessing/kamiran_calders_reweighing/test_kamiran_calders_reweighing.py]: We need a test for the correctness of the computed weights with multiple sensitive features.

Ideally, we should also have a test similar to `test_reweighing_enforces_independence` but with multiple sensitive features too.

File: fairlearn/preprocessing/_kamiran_calders_reweighing.py [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+        unique_rows, counts = np.unique(SY, axis=0, return_counts=True)
+        ps = np.array([(S == row[:-1]).all(axis=1).mean() for row in unique_rows])
+        py = np.array([(y_enc == row[-1]).mean() for row in unique_rows])
+        psy = counts / n_samples
+
+        self._weights_mapping_ = {
+            tuple(row): (p_s * p_y) / p_sy for row, p_s, p_y, p_sy in zip(unique_rows, ps, py, psy)
+        }
+
+        return self
```
Comment (taharallouche) [inline fairlearn/preprocessing/_kamiran_calders_reweighing.py]:
```suggestion
        X_valid = validate_data(self, X, ensure_2d=True, dtype=None)
        y_valid = validate_data(self, y, ensure_2d=False, dtype=None)
        n_samples = X_valid.shape[0]

        self.feature_names_in_ = (
            x.columns
            if hasattr(x := nw.from_native(X, eager_only=True, pass_through=True), "columns")
            else None
        )
        if X_valid.dtype.kind in {"O", "U", "S"}:
            self._x_encoder_ = OrdinalEncoder()
            try:
                X_enc = self._x_encoder_.fit_transform(X_valid)
            except TypeError as e:
                raise TypeError("argument must be a string or number") from e
        else:
            self._x_encoder_ = None
            X_enc = X_valid

        if y_valid.dtype.kind in {"O", "U", "S"}:
            self._y_encoder_ = OrdinalEncoder()
            try:
                y_enc = self._y_encoder_.fit_transform(y_valid.reshape(-1, 1)).ravel()
            except TypeError as e:
                raise TypeError("argument must be a string or number") from e
        else:
            self._y_encoder_ = None
            y_enc = y_valid
        self._y_array_ = y_enc

        if not sensitive_features:
            self.sensitive_idx_ = []
            self._weights_mapping_ = None
            return self

        if isinstance(sensitive_features, Iterable):
            sensitive_features = list(sensitive_features)

        self.sensitive_idx_ = []
        for f in sensitive_features:
            if isinstance(f, int):
                if f < 0 or f >= X_valid.shape[1]:
                    raise ValueError(f"Sensitive feature index {f} is out of bounds")
                self.sensitive_idx_.append(f)
            else:
                if self.feature_names_in_ is None:
                    raise ValueError(f"Sensitive feature {f} provided, but X has no column names")
                if f not in self.feature_names_in_:
                    raise ValueError(f"Sensitive feature {f} not found in columns")
                self.sensitive_idx_.append(self.feature_names_in_.index(f))

        S = X_enc[:, self.sensitive_idx_]
        SY = np.column_stack([S, y_enc])

        unique_sy, inv_sy, counts_sy = np.unique(
            SY, axis=0, return_inverse=True, return_counts=True
        )
        psy = counts_sy / n_samples

        _, inv_s, counts_s = np.unique(S, axis=0, return_inverse=True, return_counts=True)
        ps_per_sample = counts_s[inv_s] / n_samples
        ps = np.bincount(inv_sy, weights=ps_per_sample) / np.bincount(inv_sy)

        _, inv_y, counts_y = np.unique(y, return_inverse=True, return_counts=True)
        py_per_sample = counts_y[inv_y] / n_samples
        py = np.bincount(inv_sy, weights=py_per_sample) / np.bincount(inv_sy)

        self._weights_mapping_ = {
            tuple(sy): (p_s * p_y) / p_sy for sy, p_s, p_y, p_sy in zip(unique_sy, ps, py, psy)
        }

        return self
```

This is it more or less what I think it should look like with narwhals. 

PS: I also tried to vectorize the calculations of ps, py and psy. Didn't double check thoroughly for now, so we can keep your version until I do.

**Comment by taharallouche** [comment]
Regarding the API, I just saw that the maintainers had decided to use a MetaEstimator (see discussion [here](https://github.com/fairlearn/fairlearn/issues/784#issuecomment-865853776)).

You can see a past attempt using the Meta Estimator in this [PR](https://github.com/fairlearn/fairlearn/pull/1017).

It should be quite straightforward to convert your current implementation to the suggested API, as the main logic is still input validation and weight computation, I will be glad to help you with that if needed.
